services/account/libs/common/src/blockchain/blockchain.service.ts

```diff
@@ -365,13 +359,28 @@ export class BlockchainService implements OnApplicationBootstrap, OnApplicationS
         // Remove the 0x prefix from the handle and convert the hex handle to a utf-8 string
         const handleData = handleHex.slice(2);
         siwfTxnValues.handle = Buffer.from(handleData.toString(), 'hex').toString('utf-8');
+        if (!siwfTxnValues.msaId) siwfTxnValues.msaId = record.event.data.msaId.toString();
       }
       if (record.event && this.api.events.msa.DelegationGranted.is(record.event)) {
         siwfTxnValues.newProvider = record.event.data.providerId.toString();
-        const owner = record.event.data.delegatorId.toString();
+        if (!siwfTxnValues.msaId) siwfTxnValues.msaId = record.event.data.delegatorId.toString();
       }
     });
-    return siwfTxnValues as SIWFTxnValues;
+
+    // If one of the above events has previously occurred, we still need to set those values.
+    if (siwfTxnValues.handle === '') {
+      const handle = await this.getHandleForMsa(siwfTxnValues.msaId);
+      siwfTxnValues.handle = `${handle?.base_handle}.${handle?.suffix}`;
+    }
+    if (siwfTxnValues.address === '') {
+      const keyInfo = await this.getKeysByMsa(siwfTxnValues.msaId);
+      siwfTxnValues.address = keyInfo?.msa_keys[0].toString();
+    }
+    if (siwfTxnValues.newProvider === '') {
+      siwfTxnValues.newProvider = this.configService.providerId;
+    }
+
+    return siwfTxnValues;
```

Comment:
Just a note that this solution works specifically because we submit our extrinsics in a batch, so whether we submit one or multiple depending on the scenario, they're all guaranteed to be in the same block.